[Synthetics] Add monitor detail flyout

justinkambic opened the pull request:
## Summary

_Note to reviewers:_ you can test this PR using the [cloud deployment](https://kibana-pr-136156.kb.us-west2.gcp.elastic-cloud.com:9243/app/synthetics).

Credentials are available [here](https://p.elstc.co/paste/JnONQ6LV#KYQwaKSC6bI9fv-8j2lw1W01DzSLuW7e9M84raVSJ3p).

Resolves #135150.

Introduces monitor detail flyout for Synthetics UI.

![20220906164016](https://user-images.githubusercontent.com/18429259/188733840-32246a89-0814-4adc-883b-ec3dc9992d84.gif)

### Additional notes

There are a few modifications to functionality that was merged previously. I've extracted the toast notifications for the enable/disable hook to an effect that is tied to the redux action responsible for the actual API request. There were numerous cases where the current solution is vulnerable to multiple notification calls because they're inferring when the send a toast based on local app state.

This way, the notification is tied to the resolution of the request, and it handles errors as well.

## Testing

Some high-level things to test in addition to regular review scrutiny:
- Viewing and opening the flyout via the actions popover
- Opening popover by clicking metric items
- Enable/disable monitor(s) via the flyout and actions popover

justinkambic committed:
Add a monitor details flyout for the synthetics UI.

shahzad31 commented on x-pack/plugins/synthetics/public/legacy_uptime/components/monitor_management/monitor_list/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,286 @@
+  font-weight: bold;
+`;
+
+const STATUS_COLUMN_NAME = i18n.translate('xpack.synthetics.monitorList.statusColumnName', {
+  defaultMessage: 'Status',
+});
+
+const LOCATION_COLUMN_NAME = i18n.translate('xpack.synthetics.monitorList.locationColumnName', {
+  defaultMessage: 'Location',
+});
+
+const DURATION_HEADER_TEXT = i18n.translate('xpack.synthetics.monitorList.durationHeaderText', {
+  defaultMessage: 'Duration',
+});
+
+const MONITOR_DETAILS_HEADER_TEXT = i18n.translate(
+  'xpack.synthetics.monitorList.monitorDetailsHeaderText',
+  {
+    defaultMessage: 'Monitor Details',
+  }
+);
```
Please move these to the bottom of file, we have been following this pattern to keep i18n in end of file, to make component more readable.

justinkambic replied:
Sure, hadn't really intended for this to be reviewed yet 😆

shahzad31 commented on x-pack/plugins/synthetics/public/legacy_uptime/components/monitor_management/monitor_list/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,286 @@
+                    from: 'now-12h',
+                    to: 'now',
+                  },
+                  reportDefinitions: {
+                    'monitor.id': [id],
+                    'observer.geo.name': [location],
+                  },
```
```suggestion
                  reportDefinitions: {
                    'monitor.id': [id],
                  },
                  filters: [
      							{
        							field: "observer.geo.name",
        							values: [location]
      							}
    						],
```

shahzad31 replied:
it translate to better UI in exp view

dominiqueclarke committed:
synthetics - add overview grid

dominiqueclarke committed:
adjust tests and types

dominiqueclarke committed:
Update x-pack/plugins/synthetics/common/constants/client_defaults.ts

dominiqueclarke committed:
Merge branch 'main' of github.com:elastic/kibana into feature/overview-grid-list

dominiqueclarke committed:
Update x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/overview_grid.tsx

dominiqueclarke commented on x-pack/plugins/synthetics/public/legacy_uptime/components/monitor_management/monitor_list/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,286 @@
+    [id]
+  );
+  const { data: detailData, status: detailStatus } = useFetcher(() => {
+    return fetchPings({
```
I'm not sure if we should be using legacy uptime fetchPings. I believe our goal is to use the data plugin or the useEsQuery hook moving forward. Thoughts @shahzad31 ?

dominiqueclarke replied:
Also, this legacy query likely filters by `monitor.id`, whereas we'll want to filter by `config_id` moving forward.

dominiqueclarke commented on x-pack/plugins/synthetics/public/legacy_uptime/components/monitor_management/monitor_list/monitor_list.tsx:
```diff
@@ -229,6 +239,13 @@ export const MonitorManagementList = ({
+      {!!detailId && (
+        <MonitorDetailFlyout
+          id={detailId}
+          location="North America - US Central"
```
This will likely be the monitor id, not the monitor location name.

dominiqueclarke committed:
adjust formatting

dominiqueclarke committed:
implement refresh

dominiqueclarke committed:
add duration label

dominiqueclarke committed:
adjust text size of pagination

dominiqueclarke committed:
adjust store mock

dominiqueclarke committed:
Merge branch 'feature/overview-grid-list' of https://github.com/dominiqueclarke/kibana into feature/overview-grid-list

justinkambic committed:
Migrating to synthetics app. WIP.

dominiqueclarke committed:
adjust tests

dominiqueclarke committed:
adjust types

justinkambic committed:
Merge branch 'main' into feature/overview-grid-list

kibanamachine committed:
Merge branch 'main' into feature/overview-grid-list

dominiqueclarke committed:
Update x-pack/plugins/synthetics/public/apps/synthetics/hooks/use_last_x_checks.ts

justinkambic committed:
Add click handler to trigger flyout.

justinkambic committed:
Merge main.

justinkambic committed:
Revert local dev name.

justinkambic committed:
Add padding to fix layout.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Integrate flyout with overview state and implement updated design.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Change return type of api function.

justinkambic committed:
Merge main.

justinkambic committed:
Re-add location filter.

justinkambic committed:
Cleanup.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Integrate actions popover into monitor detail flyout.

justinkambic committed:
Fix i18n.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Clean up code. Only render inspect option if popover is outside of detail flyout.

justinkambic committed:
Fix types.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Fix types.

justinkambic committed:
Clean up.

justinkambic committed:
Fix status badge.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Fix some bugs, pluralize unit labels.

justinkambic committed:
Add unit tests for flyout.

justinkambic committed:
Add helper for navigating to Synthetics Overview.

justinkambic committed:
Add helper attribute to `<MetricItem />`.

justinkambic committed:
Add simple e2e test for monitor detail flyout.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Move toast creation to enabled upsert effect to prevent duplicate toast calls.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch '135150/monitor-detail-flyout' of github.com:justinkambic/kibana into 135150/monitor-detail-flyout

justinkambic committed:
Clean up types.

justinkambic committed:
Fix i18n issue.

justinkambic committed:
Trying to see if rearranging e2e tests will avoid unwanted side effect.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Extract flyout journey to dedicated test file.

justinkambic commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Only display drop shadow when `ActionPopover` button has custom positioning.

justinkambic committed:
Simplify test code.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch '135150/monitor-detail-flyout' of github.com:justinkambic/kibana into 135150/monitor-detail-flyout

justinkambic committed:
PR feedback.

justinkambic committed:
Undo immaterial file change.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

dominiqueclarke committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic commented:
Sorry to anyone who got an unnecessary ping.

justinkambic committed:
Add default values and optional props for flyout duration viz lookback window.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Mergh origin branch 'main' into 135150/monitor-detail-flyout

dominiqueclarke reviewed (changes requested):
![image](https://user-images.githubusercontent.com/11356435/191578531-9171ebd1-28d3-471c-b9d0-f3cb3f3df640.png)

Tags also are stretched
<img width="853" alt="Screen Shot 2022-09-21 at 2 08 37 PM" src="https://user-images.githubusercontent.com/11356435/191579034-4a9f2ba8-437c-456b-8a8f-9551e48570d2.png">

justinkambic committed:
Fix missing `onClose`.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge main.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

hbharding commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+  );
+}
+
+function BodyInfo({ header, content }: { header: string; content: JSX.Element | string }) {
```
The markup of this component should be using `EuiDescriptionList` and `EuiDescriptionListItem`s. Use `type="column"` and `compressed={true}` and `gutterSize="s"` for description list.

justinkambic replied:
We're currently blocked for the `gutterSize` prop [until EUI 66 lands in Kibana](https://github.com/elastic/kibana/pull/141279). I'll keep a pin in this and address if we end up still not being able to set this prop.

hbharding commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+              header={URL_HEADER_TEXT}
+              content={
+                monitorDetail.data?.url?.full ? (
+                  <EuiLink href={monitorDetail.data.url.full}>
```
Add `external` or `target="_blank"` to `EuiLink` so that the "popOut" icon appears -> this helps indicate to user that this will open a page in a new tab.

justinkambic replied:
9c8083c5925161e6703e28178304e3d707a2edd3

hbharding reviewed (changes requested):
Thanks @dominiqueclarke and @justinkambic 

I only reviewed the markup for the EuiFlyout (using Dominique's screenshot as a reference). I believe I marked everything she found and provided comments about which props to use or guidance on proper component usage.

I'd be happy to hop on a zoom to do a quick review when you're ready, or just share screenshots when you're done and I can review async. Either works. Thanks!

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+                  name: LOCATION_COLUMN_NAME,
+                  field: 'observer.geo.name',
+                  render: () => (
+                    <EuiSelect
```
Thanks for this!

justinkambic committed:
Provide appropriate size/type props for flyout.

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+  const locationStatuses = useStatusByLocation(id);
+  const locations = locationStatuses.locations?.filter((l: any) => !!l?.observer?.geo?.name) ?? [];
+  return (
+    <EuiFlyout onClose={props.onClose}>
```
37314de62201daa531067023500b95a81f73dc39

justinkambic committed:
Capitalize dynamic copy.

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+                  field: 'monitor.status',
+                  render: (statusString: string) => (
+                    <EuiBadge color={statusString === 'up' ? 'success' : 'danger'}>
+                      {statusString}
```
72f8b4f0dcd91e8e228331848da629c6aef835ea

justinkambic committed:
Add appropriate `size` values to elements.

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+            <EuiTitle size="s">
+              <h3>{MONITOR_DETAILS_HEADER_TEXT}</h3>
+            </EuiTitle>
+            <EuiSpacer />
```
676ec3148f142ea12e10c86c65d61efd827d0bc9

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+          <EuiFlyoutHeader hasBorder>
+            <EuiFlexGroup>
+              <EuiFlexItem grow={false}>
+                <EuiTitle>
```
676ec3148f142ea12e10c86c65d61efd827d0bc9

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+              ]}
+            />
+            <EuiSpacer />
+            <EuiTitle size="s">
```
676ec3148f142ea12e10c86c65d61efd827d0bc9

justinkambic commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitors_page/overview/overview/monitor_detail_flyout.tsx:
```diff
@@ -0,0 +1,460 @@
+            />
+          </EuiFlyoutHeader>
+          <EuiFlyoutBody>
+            <EuiTitle size="s">
```
676ec3148f142ea12e10c86c65d61efd827d0bc9

justinkambic committed:
Add `external` prop to monitor URL anchor.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
TEMP.

justinkambic committed:
Add context menu to location select, fix layouts.

justinkambic committed:
Delete unused files.

justinkambic committed:
Break chart into separate component.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Fix broken test.

justinkambic committed:
Add `aria-label` for new icon button.

justinkambic committed:
Saving some dev stuff to test on cloud.

justinkambic committed:
Add translations for up/down values.

justinkambic committed:
More dev testing for cloud.

justinkambic committed:
Make monitor detail refresh on location change as well.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Remove console logs.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Fix bug where clicking between overview items would not update flyout details.

justinkambic committed:
Simplify typing of action creator.

justinkambic committed:
Switch exploratory view chart to average operation.

justinkambic committed:
Fix types.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Implement designer feedback.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Improve formatting, translation.

dominiqueclarke commented:
@elasticmachine merge upstream

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Make `go to monitor` button regular, i.e. not filled.

justinkambic committed:
Fix copy for close flyout button.

justinkambic committed:
Fix broken unit test.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

dominiqueclarke reviewed (commented):
<img width="418" alt="Screen Shot 2022-10-25 at 7 20 06 PM" src="https://user-images.githubusercontent.com/11356435/197900139-884fd9e9-51ab-4d50-9b55-fbf906e4c8f5.png">
Duration chart is not rendered for project monitors, but honestly I'm considering deferring this fix until https://github.com/elastic/kibana/issues/143309

![image](https://user-images.githubusercontent.com/11356435/197900267-25b5db5c-17a0-4127-9c5a-f8cf562cdaa8.png)

Tags still look a little off
![image](https://user-images.githubusercontent.com/11356435/197900335-f858ba4c-90c9-47a3-95c1-f2f34dd4b7ae.png)

This should be HTTP, TCP, ICMP, Multistep, or Single Page I believe. The value is available at `attributes[ConfigKey.FORM_MONITOR_TYPE]`

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic committed:
Prefer `ConfigKey` constants for attribute names.

justinkambic committed:
Prefer to use `EuiBadgeGroup` for tags.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

justinkambic commented:
@dominiqueclarke I think I have addressed all the feedback from your last review except for the issue with Project Monitors. If you'd like, we can add this to the list of things we plan to iteratively fix. I'm going to make a follow-up issue detailing items we are blocked on or delaying for various reasons.

justinkambic committed:
Merge branch 'main' into 135150/monitor-detail-flyout

shahzad31 committed:
Merge branch 'main' into 135150/monitor-detail-flyout

shahzad31 committed:
Merge branch 'main' of github.com:elastic/kibana into 135150/monitor-detail-flyout

shahzad31 committed:
fix type

dominiqueclarke reviewed (approved):
LGTM

shahzad31 committed:
Merge branch 'main' of github.com:elastic/kibana into 135150/monitor-detail-flyout

shahzad31 committed:
fix conflicts

hbharding reviewed (approved):
lgtm thanks @justinkambic
